onload/deviceplugin/1000-deviceplugin.yaml

```diff
@@ -65,7 +65,13 @@ spec:
         lifecycle:
           postStart:
             exec:
-              command: ["cp", "-TRv", "/opt/onload", "/host/onload"]
+              command:
+              - /bin/sh
+              - -c
+              - |
+                set -e;
+                cp -TRv /opt/onload /host/onload;
+                chcon --verbose --type container_file_t --recursive /host/onload/;
```

Comment:
Thanks for the comprehensive comment.

> How about `cp --context=container_file_t ...`

Thanks, I only used `--preserve=context` in the past. However, it doesn't do what we want from within a device plugin pod:
```console
sh-4.4# cp -v --context='system_u:object_r:container_file_t:s0' /host/onload/foo.txt /host/onload/bar.txt
cp: warning: ignoring --context; it requires an SELinux-enabled kernel
'/host/onload/foo.txt' -> '/host/onload/bar.txt'
```

It didn't apply the SELinux labels, and to understand why I ran it with `strace`:
```console
# strace cp -v --context='system_u:object_r:container_file_t:s0' /host/onload/foo.txt /host/onload/bar.txt
...
access("/etc/selinux/config", F_OK)     = -1 ENOENT (No such file or directory)
prctl(PR_SET_NAME, "cp")                = 0
prctl(PR_SET_KEEPCAPS, 140726852443438) = -1 EINVAL (Invalid argument)
geteuid()                               = 0
write(2, "cp: ", 4cp: )                     = 4
write(2, "warning: ignoring --context; it "..., 66warning: ignoring --context; it requires an SELinux-enabled kernel) = 66
```

It looks like our base image is not SELinux-friendly. But I'm not keen to choose another image only for that purpose!

> A slightly more persistent version would be

Yes, and I tried to explain that in the commit/PR message: it implies changes to the host file system, which we probably should do with MC rather than from within a pod, and it seems too heavy-weight for this occasion.

> a simple initContainer for the file transfer

IMO, this is also heavy-weight because we would have to "mount" another image merely for a couple of shell commands.

> - an ./entrypoint.sh for the device plugin container, or
> - performing the file transfer within device plugin Go code itself.

Interesting. Either of these two options is better than `postStart` alone, at least because they would keep the copying logs. (The lifecycle hooks are dull, exposing the logs only on failures in the form of an event. 🤦‍♂️ ) However, this made me realise that we probably want to use the `preStop` to remove the Onload files, and it is somewhat more reliable than the entry point or Golang because it's handled with K8s itself.

> OCP v4.12 has first-class [definitions](https://docs.openshift.com/container-platform/4.12/security/security_profiles_operator/spo-selinux.html) in the form of `SelinuxProfile`.

Do you have insights about that kind, i.e. how to harness it for our use case (copying files to the host FS)?

Reply:
TL;DR: I will add the `preStop` lifecycle hook to remove host files, but please comment if you disagree.

Reply:
"it requires an SELinux-enabled kernel" is a message from [coreutils cp](https://github.com/coreutils/coreutils/blob/master/src/cp.c#L1170), conditional on the value set by `is_selinux_enabled ()`, a function provided by `libselinux` package. Perhaps that package is missing from the container and it has defaulted to assuming disabled. That package [is in ubi repos](https://catalog.redhat.com/software/containers/ubi8/ubi/5ff3f50a831939b08d1b832a?container-tabs=packages&tag=8.7-8&push_date=1680612950000).

For reference, `libselinux-devel` package contains: `/usr/include/selinux/selinux.h` & `/usr/share/man/man3/is_selinux_enabled.3.gz`.

---

>> a simple initContainer for the file transfer
>
> IMO, this is also heavy-weight because we would have to "mount" another image

The onload-user image is already consumed on the node, and this would be a per-device-plugin container not per-user-pod, so very low cost. Hopefully avoidable anyway.

---

Details of applying `SelinuxProfile` are in the link, under "Applying SELinux profiles to a pod", however we cannot explore this while v4.10 is considered supported. Ideally we should define a custom SELinux Context (eg. `onload_container_file_t`) so the user pod is only exposed to our files.

Reply:
Thanks again.

I did a bit of more research on the topic. Apparently, `libselinux` is already installed with the base:
```console
sh-4.4# rpm -q libselinux
libselinux-2.9-8.el8.x86_64
```

In the `strace` report above, `cp` was complaining about missing `/etc/selinux/config`, so I checked where it comes from:
```console
$ yum whatprovides '/etc/selinux/config' | grep --before-context=3 Filename | head -4
selinux-policy-3.14.1-61.el8.noarch : SELinux policy configuration
Repo        : lab-rhel8.6-x86_64-baseos
Matched from:
Filename    : /etc/selinux/config
```

Then, installed `selinux-policy` in a pod:
```
sh-4.4# microdnf install selinux-policy
```

Retried `cp` still "fails" (exit code is 0):
```console
sh-4.4# cp -v --context='system_u:object_r:container_file_t:s0' /host/onload/foo.txt /host/onload/bar.txt
cp: warning: ignoring --context; it requires an SELinux-enabled kernel
'/host/onload/foo.txt' -> '/host/onload/bar.txt'
```

However, now the `strace` report is slightly different:
```console
sh-4.4# strace cp -v --context='system_u:object_r:container_file_t:s0' /host/onload/foo.txt /host/onload/bar.txt
access("/etc/selinux/config", F_OK)     = 0
prctl(PR_SET_NAME, "cp")                = 0
prctl(PR_SET_KEEPCAPS, 140721982192946) = -1 EINVAL (Invalid argument)
geteuid()                               = 0
write(2, "cp: ", 4cp: )                     = 4
write(2, "warning: ignoring --context; it "..., 66warning: ignoring --context; it requires an SELinux-enabled kernel) = 66
write(2, "\n", 1
)                       = 1
```
I traced a little further with `gdb` and discovered that `is_selinux_enabled()` is checking `has_selinux_config`, which is True, and `selinux_mnt`, which is False. The `selinux_mnt` is set in `init_selinuxmnt()` which is checking that:
- `/sys/fs/selinux` (`SELINUXMNT`) is present in a pod but has a wrong magic value. The `f_type` returned with `statfs` is `0x62656572` (`SYSFS_MAGIC`) while expected to be `0xf97cff8c` (`SELINUX_MAGIC`). https://man7.org/linux/man-pages/man2/statfs.2.html
- `/selinux` (`OLDSELINUXMNT`) is not present.
- `/proc/mounts` (legacy way of discovering SELinux) does not contain a line with SELinux.

I tried to set `shareProcessNamespace`, `hostIPC` and `hostPID` out of curiosity if it helps to peek into the node's procfs, but it doesn't achieve that goal. (I don't want to escape the namespace merely to make the pod SELinux-friendly even if it worked.)

Another thought is that, presumably, the userland SELinux library must be compatible with the kernel's SELinux, and not any base image would be a good fit for the purpose, limiting our choice for the device plugin, which is meant to be no more than a Golang runtime.

Overall, I think `chcon` is the way to go for this exercise.

Re on `initContainer`: I understand it is a low overhead, but I don't see why it's better in the light of `preStop`?

Re on `SelinuxProfile`: ACK. Overall, it looks interesting as it makes SELinux more kube-friendly.